### PR TITLE
Add support for native ETCD distributed locks

### PR DESCRIPTION
**Pull request by dcepelik**
Hi Shaun,

as promised.

One new test case is commented out as it's failing, this is because you have a sophisticated helper in place which I wanted to leverage but which cannot be used for the `lock` gRPC call (obviously it will time out, because it's locked from the previous call in the same helper call). I don't know the proper Ruby way of making this work (unlocking in the helper), so I was hoping you would have the capacity to fix that one test case, or at least provide some hints how to do it.

I've done extensive local testing and everything seems to work well. The semantics of the locking/unlocking are really simple, as the RPC call will fail, or a time-out will occur, or the locking/unlocking has succeeded. Still this deserves a thorough examination, because I'm not much of a Ruby programmer and have been fighting with the tools. (For example, I did not find a way to convince gRPC compiler to generate relative requires, so I had to fix them manually.)

One more note, most pieces of protobuf code were taken from the official ETCD github repo, but it's not mentioned anywhere. (But I suspect that's the case for other protobuf code in the gem as well.)

Please let me know if I can do more.

David

### CODE REVIEW

**Comment by graywolf-at-work** [comment]
If this is merged, #115 can be closed since this fixes the same issue.

File: README.md [hunk 1]
```diff
@@ -164,6 +164,20 @@ conn.watch('boom') do |events|
 end
 ```
 
+## Locks
+```ruby
+# Attempt to lock distibuted lock 'foo', wait at most 10 seconds
+lock_key = conn.lock('foo', timeout: 10).key
```
Comment (davissp14) [inline README.md]: Looks like this example is missing the lease requirement.

Comment (davissp14) [inline README.md]: Looks like this example is missing the lease requirement.

File: README.md [hunk 2]
```diff
@@ -164,6 +164,20 @@ conn.watch('boom') do |events|
...
+
+# Unlock the 'foo' lock using the key returned from `lock`
+conn.unlock(key)
+
+# Perform a critical section while holding the lock 'hello'
+conn.with_lock('hello') do
```
Comment (davissp14) [inline README.md]: This also requires a lease, right?

File: lib/etcdv3/auth.rb [hunk 3]
```diff
@@ -122,7 +123,7 @@ def generate_token(user, password, timeout: nil)
     private
 
     def deadline(timeout)
-      Time.now.to_f + (timeout || @timeout)
+      from_relative_time(timeout || @timeout)
```
Comment (davissp14) [inline lib/etcdv3/auth.rb]: Nice find!

**Comment by dcepelik** [comment]
Fixed.

**Comment by davissp14** [comment]
Nice work!

**Pull request merged**